### PR TITLE
fix recv message callback request & response decision

### PR DESCRIPTION
**Pull request by imzhenyu**

### CODE REVIEW

**Comment by imzhenyu** [comment]
@qinzuoyan please help review.

**Comment by qinzuoyan** [comment]
ok~

File: include/dsn/internal/network.h [hunk 1]
```diff
@@ -113,7 +113,7 @@ namespace dsn {
         //
         //
         //
-        void on_recv_reply(message_ex* msg, int delay_ms);
+        void on_recv_reply(uint64_t id, message_ex* msg, int delay_ms);
```
Comment (qinzuoyan) [inline include/dsn/internal/network.h:116]: why need we pass param id here?

File: src/core/core/network.cpp [hunk 2]
```diff
@@ -367,32 +369,22 @@ namespace dsn
         return true;
     }
 
-    bool rpc_session::on_recv_reply(uint64_t key, message_ex* reply, int delay_ms)
+    void rpc_session::on_recv_message(message_ex* msg, int delay_ms)
```
Comment (qinzuoyan) [inline src/core/core/network.cpp:372]: reserve the left comments here?

File: include/dsn/internal/network.h (hunk 1 above)
Comment (imzhenyu) [inline include/dsn/internal/network.h:116]: due to the fast failure we have when connect failed, i.e., `on_recv_reply(request->header->id, nullptr, 0);`

File: src/core/core/network.cpp (hunk 2 above)
Comment (imzhenyu) [inline src/core/core/network.cpp:372]: np.

File: include/dsn/internal/network.h (hunk 1 above)
Comment (qinzuoyan) [inline include/dsn/internal/network.h:116]: Yes, I see

Comment (qinzuoyan) [inline include/dsn/internal/network.h:116]: And, maybe we'd better add some function comments here

**Comment by qinzuoyan** [comment]
I suggest combining ERR_NETWORK_INIT_FAILED and ERR_NETWORK_START_FAILED to uniform  ERR_NETWORK_FAILURE, and use it for all network failure.

Specially, in rpc_client_matcher::on_recv_reply(), we'd better set err to ERR_NETWORK_FAILURE when reply == nullptr, to differ from ERR_TIMEOUT, which I think is meaningful for user, and also, there is only two kinds of err code (ERR_TIMEOUT, ERR_NETWORK_FAILURE), it's easy to use.

How do you think about it?

**Comment by imzhenyu** [comment]
It is reasonable. However, our current application-level code (e.g., in replication) may not handle that failure appropriately yet - need to look into the error-code warnings (i.e., "not handled") carefully after the change.

**Comment by qinzuoyan** [comment]
sure, let's leave it to future, you can merge it now :)

**Pull request merged**